static/js/components/UploadSpectrum.jsx

```diff
@@ -0,0 +1,494 @@
+          const telescope = telescopes.find(
+            (t) => t.id === instrument.telescope_id
+          );
+          let name = "";
+          if (telescope) {
+            name += `${telescope.nickname} / `;
+          }
+          name += instrument.name;
+          return name;
+        }),
+      },
+      wave_column: {
+        type: "integer",
+        default: 0,
+        title: "0-Based Wavelength Column Index",
+      },
+      flux_column: {
+        type: "integer",
+        default: 1,
+        title: "0-Based Flux Column Index",
+      },
+      has_fluxerr: {
+        type: "string",
+        default: "No",
+        title: "Does your spectrum have flux errors?",
+        enum: ["No", "Yes"],
+      },
+    },
+    required: [
+      "has_fluxerr",
+      "mjd",
+      "wave_column",
+      "flux_column",
+      "instrument_id",
+    ],
+    dependencies: {
+      has_fluxerr: {
+        oneOf: [
+          {
+            properties: {
+              has_fluxerr: {
+                enum: ["No"],
+              },
+            },
+          },
+          {
+            properties: {
+              has_fluxerr: {
+                enum: ["Yes"],
+              },
+              fluxerr_column: {
+                type: "integer",
+                default: null,
+                title: "0-Based Flux Error Column Index",
+              },
+            },
+            required: ["has_fluxerr", "fluxerr_column"],
+          },
+        ],
+      },
+    },
+  };
+
+  const parseAscii = ({ formData }) => {
+    dispatch({ type: Actions.RESET_PARSED_SPECTRUM });
+    const ascii = dataUriToBuffer(formData.file).toString();
+    const payload = {
+      ascii,
+      flux_column: formData.flux_column,
+      wave_column: formData.wave_column,
+      fluxerr_column:
+        formData?.has_fluxerr === "Yes" ? formData.fluxerr_column : null,
+    };
+    dispatch(Actions.parseASCIISpectrum(payload));
+  };
+
+  const uploadSpectrum = async () => {
+    if (!parsed) {
+      throw new Error("No spectrum loaded on frontend.");
+    }
+
+    const ascii = dataUriToBuffer(persistentFormData.file).toString();
+    const filename = persistentFormData.file.split(";")[1].split("name=")[1];
+    const payload = {
+      ascii,
+      flux_column: persistentFormData.flux_column,
+      wave_column: persistentFormData.wave_column,
+      fluxerr_column:
+        persistentFormData?.has_fluxerr === "Yes"
+          ? persistentFormData.fluxerr_column
+          : null,
+      obj_id: route.id,
+      instrument_id: persistentFormData.instrument_id,
+      observed_at: dayjs
+        .unix((persistentFormData.mjd - 40_587) * 86400)
+        .utc()
+        .format(),
+      filename,
+    };
+    const result = await dispatch(Actions.uploadSpectrumAscii(payload));
+    if (result.status === "success") {
+      dispatch(showNotification("Upload successful."));
+      dispatch({ type: RESET_PARSED_SPECTRUM });
+      setPersistentFormData({
```

Comment:
If you pull out `defaultFormData` from the useEffect above to be in component-wide scope, you can re-use it here rather than duplicating this obj literal.

Reply:
Can't do it as the two are slightly different. The one in `useEffect` depends for the result of a blocking function, whereas this one uses redux state.